### PR TITLE
[fix #13231] sigmoid : add OpenCL support

### LINKED ISSUE #13231: sigmoid : add OpenCL support
All said in the title, this has been discussed and would be nice to have in 4.4 as many users are now using Sigmoid instead of Filmic. Even if this module is fast having some more CPU cycle saved would be nice.

### PR DESCRIPTION
**Pull request by erzsig**
[fix #13231] sigmoid : add OpenCL support
I have adapted the code from sigmoid.c process function to opencl. As far as I can see the modul now has the same effect on the image either opencl is switched on or off.

### CODE REVIEW

File: data/kernels/sigmoid.cl [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+static inline float4 desaturate_negative_values(const float4 i)
+{
+  const float pixel_average = fmax((i.x + i.y + i.z) / 3.0f, 0.0f);
+  const float min_value = fmin(fmin(i.x, i.y), i.z);
+  const float saturation_factor = min_value < 0.0f ? -pixel_average / (min_value - pixel_average) : 1.0f;
+
+  return pixel_average + saturation_factor * (i - pixel_average);
+}
+
+static inline float generalized_loglogistic_sigmoid_scalar(const float value, const float magnitude, const float paper_exp,
```
Comment (TurboGit) [inline data/kernels/sigmoid.cl:90]: some style changes to follow dt convention.

- static routines must start with underscore
- the parameters one per line when there is too many to fit on a line

So it should be:

```
static inline float _generalized_loglogistic_sigmoid_scalar(const float value, 
                                                            const float magnitude, 
                                                            const float paper_exp,
                                                            const float film_fog, 
                                                            const float film_power, 
                                                            const float paper_power)
{
```

Many other occurrences in following code.

File: data/kernels/sigmoid.cl [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    per_channel[order.mid] - per_channel[order.min] < 1e-9f )
+  {
+    pix_io[0] = per_channel[0];
+    pix_io[1] = per_channel[1];
+    pix_io[2] = per_channel[2];
+    return;  // Nothing to fix
+  }
+
+  // Naive Hue correction of the middle channel
+  const float full_hue_correction = per_channel[order.min] + ((per_channel[order.max] - per_channel[order.min]) * (pix_io[order.mid] - pix_io[order.min]) / (pix_io[order.max] - pix_io[order.min]));
```
Comment (TurboGit) [inline data/kernels/sigmoid.cl:138]: line too long, please split for readability

**Review by TurboGit — CHANGES_REQUESTED** [review]
Some style comments, not tested yet.

**Comment by TurboGit** [comment]
Just tested this implementation and to me it works fine. The integration test 0097 pass fine.

**Comment by erzsig** [comment]
Hi @TurboGit , thanks for your comments and the test.
I have corrected the style, and checked that the code remained functional.
Please, review it again.

**Review by TurboGit — APPROVED** [review]
Thanks!

**Comment by MStraeten** [comment]
kernel doesn't compile with macos arm64:
```
[opencl_synchronization_timeout] synchronization timeout set to 200
     1,1256 [opencl_create_kernel] could not create kernel `sigmoid_loglogistic_per_channel'! (CL_INVALID_PROGRAM)
     1,1256 [opencl_create_kernel] could not create kernel `sigmoid_loglogistic_per_channel_interpolated'! (CL_INVALID_PROGRAM)
     1,1257 [opencl_create_kernel] could not create kernel `sigmoid_loglogistic_rgb_ratio'! (CL_INVALID_PROGRAM)
```
all other kernels compiles fine

**Comment by MStraeten** [comment]
one macos update later it's fine again ;)

**Pull request merged**